File: agent-runtime/test/transcriber.test.ts
import assert from "node:assert/strict"
import { test } from "node:test"

import type { AudioFrame, AudioSource } from "../src/media/types.js"
import { SpeechTranscriber } from "../src/speech/transcriber.js"
import type {
  SttEvent,
  StreamingSttProvider,
  StreamingSttSession,
} from "../src/speech/types.js"

function audio(value: number): AudioFrame {
  return {
    codec: "opus",
    sampleRateHz: 48_000,
    channels: 2,
    timestampMs: value,
    data: new Uint8Array([value]),
  }
}

function source(id: string, trackName = "mic"): AudioSource {
  return { participantId: id, participantName: id, trackName }
}

class ControlledSession implements StreamingSttSession {
  readonly frames: AudioFrame[] = []
  closed = false
  constructor(private readonly finalEvent?: SttEvent) {}
  private readonly output: SttEvent[] = []
  private readonly waiters: ((result: IteratorResult<SttEvent>) => void)[] = []

  async pushAudio(frame: AudioFrame): Promise<void> {
    this.frames.push(frame)
  }

  emit(event: SttEvent): void {
    const waiter = this.waiters.shift()
    if (waiter) waiter({ value: event, done: false })
    else this.output.push(event)
  }

  async *events(): AsyncIterable<SttEvent> {
    while (!this.closed) {
      if (this.output.length > 0) {
        yield this.output.shift()!
        continue
      }
      const next = await new Promise<IteratorResult<SttEvent>>((resolve) =>
        this.waiters.push(resolve)
      )
      if (next.done) return
      yield next.value
    }
  }

  async close(): Promise<void> {
    if (this.finalEvent) this.emit(this.finalEvent)
    this.closed = true
    while (this.waiters.length > 0)
      this.waiters.shift()!({ value: undefined, done: true })
  }
}

async function waitFor(
  predicate: () => boolean,
  timeoutMs = 1_000
): Promise<void> {
  const start = Date.now()
  while (!predicate()) {
    if (Date.now() - start > timeoutMs) throw new Error("test timed out")
    await new Promise((resolve) => setTimeout(resolve, 0))
  }
}

test("transcriber creates one attributed session per Human track", async () => {
  const sessions: ControlledSession[] = []
  const provider: StreamingSttProvider = {
    async createSession() {
      const session = new ControlledSession()
      sessions.push(session)
      return session
    },
  }
  const events: { source: AudioSource; event: SttEvent }[] = []
  const transcriber = new SpeechTranscriber({
    provider,
    onEvent: (event) => events.push(event),
  })
  transcriber.acceptAudio(source("human-a"), audio(1))
  transcriber.acceptAudio(source("human-b"), audio(2))
  await waitFor(() => sessions.length === 2)
  assert.deepEqual(
    sessions.map((session) => session.frames[0]?.data[0]),
    [1, 2]
  )

  sessions[0]!.emit({ type: "committed", text: "A" })
  await waitFor(() => events.length === 1)
  assert.equal(events[0]!.source.participantId, "human-a")
  assert.equal(events[0]!.event.type, "committed")

  transcriber.handleMediaEvent({
    type: "audioTrackEnded",
    participantId: "human-a",
    trackName: "mic",
    reason: "participant_left",
  })
  await waitFor(() => sessions[0]!.closed)
  await transcriber.close()
  assert.equal(sessions[1]!.closed, true)
})

test("transcriber fails only the overflowing track and keeps other tracks alive", async () => {
  const sessions: ControlledSession[] = []
  const provider: StreamingSttProvider = {
    async createSession() {
      await new Promise((resolve) => setTimeout(resolve, 10))
      const session = new ControlledSession()
      sessions.push(session)
      return session
    },
  }
  const events: { source: AudioSource; event: SttEvent }[] = []
  const transcriber = new SpeechTranscriber({
    provider,
    maxFramesPerTrack: 2,
    onEvent: (event) => events.push(event),
  })
  transcriber.acceptAudio(source("overflow"), audio(1))
  transcriber.acceptAudio(source("overflow"), audio(2))
  transcriber.acceptAudio(source("overflow"), audio(3))
  transcriber.acceptAudio(source("healthy"), audio(4))
  await waitFor(() => events.some((event) => event.event.type === "error"))
  const error = events.find((event) => event.event.type === "error")!
  assert.equal(error.source.participantId, "overflow")
  await waitFor(() => sessions.length === 2)
  assert.equal(sessions[0]!.closed, true)
  assert.equal(sessions[1]!.frames[0]?.data[0], 4)
  await transcriber.close()
})

test("transcriber keeps bounded startup audio while the provider handshakes", async () => {
  const sessions: ControlledSession[] = []
  let releaseSession!: () => void
  const sessionReady = new Promise<void>((resolve) => {
    releaseSession = resolve
  })
  const events: { source: AudioSource; event: SttEvent }[] = []
  const provider: StreamingSttProvider = {
    async createSession() {
      await sessionReady
      const session = new ControlledSession()
      sessions.push(session)
      return session
    },
  }
  const transcriber = new SpeechTranscriber({
    provider,
    onEvent: (event) => events.push(event),
  })
  for (let value = 1; value <= 64; value += 1)
    transcriber.acceptAudio(source("handshake"), audio(value))

  assert.equal(
    events.some((event) => event.event.type === "error"),
    false
  )
  releaseSession()
  await waitFor(() => sessions.length === 1)
  await waitFor(() => sessions[0]!.frames.length === 64)
  assert.equal(sessions[0]!.frames[0]!.data[0], 1)
  assert.equal(sessions[0]!.frames[63]!.data[0], 64)
  await transcriber.close()
})

test("transcriber drains a provider's final event before ending a track", async () => {
  let session: ControlledSession | undefined
  const events: { source: AudioSource; event: SttEvent }[] = []
  const provider: StreamingSttProvider = {
    async createSession() {
      const created = new ControlledSession({
        type: "committed",
        text: "最后一句",
      })
      session = created
      return created
    },
  }
  const transcriber = new SpeechTranscriber({
    provider,
    onEvent: (event) => events.push(event),
  })
  transcriber.acceptAudio(source("human"), audio(1))
  await waitFor(() => Boolean(session))

  transcriber.handleMediaEvent({
    type: "audioTrackEnded",
    participantId: "human",
    trackName: "mic",
    reason: "participant_left",
  })
  await waitFor(() => Boolean(session?.closed && events.length === 1))
  assert.equal(events[0]!.event.type, "committed")
  assert.equal(
    events[0]!.event.type === "committed" ? events[0]!.event.text : undefined,
    "最后一句"
  )
  await transcriber.close()
})

File: agent/internal/speech/transcriber_test.go
package speech

import (
	"sync"
	"testing"
	"time"
)

// fakeSttSession implements StreamingSttSession with scripted events.
type fakeSttSession struct {
	events chan SttEvent
	closed chan struct{}
	once   sync.Once
}

func newFakeSttSession() *fakeSttSession {
	return &fakeSttSession{events: make(chan SttEvent, 16), closed: make(chan struct{})}
}

func (f *fakeSttSession) PushAudio(frame AudioFrame) error {
	select {
	case <-f.closed:
		return nil
	default:
	}
	return nil
}

func (f *fakeSttSession) Events() <-chan SttEvent { return f.events }

func (f *fakeSttSession) emit(event SttEvent) {
	select {
	case f.events <- event:
	default:
	}
}

func (f *fakeSttSession) Close() error {
	f.once.Do(func() { close(f.closed) })
	return nil
}

// fakeSttProvider tracks per-payload sessions so tests can attribute
// emissions deterministically regardless of goroutine scheduling.
type fakeSttProvider struct {
	mu        sync.Mutex
	sessions  []*fakeSttSession
	byPayload map[string]*fakeSttSession
}

func newFakeSttProvider() *fakeSttProvider {
	return &fakeSttProvider{byPayload: make(map[string]*fakeSttSession)}
}

func (p *fakeSttProvider) CreateSession(frame AudioFrame) (StreamingSttSession, error) {
	session := newFakeSttSession()
	p.mu.Lock()
	p.sessions = append(p.sessions, session)
	p.byPayload[string(frame.Data)] = session
	p.mu.Unlock()
	return session, nil
}

func (p *fakeSttProvider) sessionFor(payload []byte) *fakeSttSession {
	p.mu.Lock()
	defer p.mu.Unlock()
	return p.byPayload[string(payload)]
}

func opusFrame() AudioFrame {
	return AudioFrame{Codec: "opus", SampleRateHz: 48000, Channels: 2, Data: []byte{0x1, 0x2}}
}

func TestTranscriberAttributionAndCommittedOnly(t *testing.T) {
	provider := newFakeSttProvider()
	var mu sync.Mutex
	var received []AttributedSttEvent
	transcriber := NewTranscriber(provider, func(event AttributedSttEvent) {
		mu.Lock()
		received = append(received, event)
		mu.Unlock()
	})
	alice := AudioSource{ParticipantID: "human-1", ParticipantName: "Alice", TrackName: "mic"}
	bob := AudioSource{ParticipantID: "human-2", ParticipantName: "Bob", TrackName: "mic"}

	aliceFrame := opusFrame()
	bobFrame := opusFrame()
	bobFrame.Data = []byte{0x2, 0x3}
	transcriber.AcceptAudio(alice, aliceFrame)
	transcriber.AcceptAudio(bob, bobFrame)
	deadline := time.Now().Add(2 * time.Second)
	for provider.sessionFor(aliceFrame.Data) == nil || provider.sessionFor(bobFrame.Data) == nil {
		if time.Now().After(deadline) {
			t.Fatal("one session per speaker never appeared")
		}
		time.Sleep(5 * time.Millisecond)
	}
	provider.sessionFor(aliceFrame.Data).emit(SttEvent{Type: "committed", Text: "alice says hi"})
	provider.sessionFor(bobFrame.Data).emit(SttEvent{Type: "committed", Text: "bob says hey"})
	time.Sleep(50 * time.Millisecond)

	mu.Lock()
	count := len(received)
	mu.Unlock()
	if count != 2 {
		t.Fatalf("committed events lost: %d", count)
	}
	mu.Lock()
	copyReceived := append([]AttributedSttEvent(nil), received...)
	mu.Unlock()
	bySpeaker := map[string]string{}
	for _, event := range copyReceived {
		if event.Event.Type != "committed" {
			t.Fatalf("non-committed leaked: %+v", event.Event)
		}
		bySpeaker[event.Source.ParticipantName] = event.Event.Text
	}
	if bySpeaker["Alice"] != "alice says hi" || bySpeaker["Bob"] != "bob says hey" {
		t.Fatalf("attribution mismatch: %v", bySpeaker)
	}

	// TrackEnded cleans the pipeline without error.
	transcriber.TrackEnded(alice)
	transcriber.Close()
}

func TestTranscriberQueueOverflowFailsTrackSafely(t *testing.T) {
	provider := newFakeSttProvider()
	transcriber := NewTranscriber(provider, func(AttributedSttEvent) {})
	source := AudioSource{ParticipantID: "h", ParticipantName: "H", TrackName: "mic"}
	// 200 frames > MaxFramesPerTrack(128) with an unreadable provider pump
	// (the fake session accepts instantly, so this exercises the queue cap
	// on the accept path only when the pump is delayed — use a blocking
	// provider instead).
	blocking := &blockingProvider{err: nil}
	transcriber2 := NewTranscriber(blocking, func(event AttributedSttEvent) {
		if event.Event.Type != "error" {
			t.Fatalf("expected only error events, got %+v", event.Event)
		}
	})
	for i := 0; i < MaxFramesPerTrack+5; i++ {
		transcriber2.AcceptAudio(source, opusFrame())
	}
	// The first pump blocks in CreateSession; the queue must have failed the
	// track (error emitted) rather than growing unbounded.
	time.Sleep(50 * time.Millisecond)
	_ = transcriber
	transcriber.Close()
	transcriber2.Close()
}

// blockingProvider never returns a session.
type blockingProvider struct {
	err error
}

func (b *blockingProvider) CreateSession(frame AudioFrame) (StreamingSttSession, error) {
	if b.err != nil {
		return nil, b.err
	}
	select {} // blocks forever; the test's overflow path fires first
}

func TestProviderFailureFailsTrackWithoutKillingOthers(t *testing.T) {
	provider := newFakeSttProvider()
	transcriber := NewTranscriber(provider, func(AttributedSttEvent) {})
	alice := AudioSource{ParticipantID: "a", ParticipantName: "Alice", TrackName: "mic"}
	bob := AudioSource{ParticipantID: "b", ParticipantName: "Bob", TrackName: "mic"}
	transcriber.AcceptAudio(alice, opusFrame())
	transcriber.AcceptAudio(bob, opusFrame())
	time.Sleep(100 * time.Millisecond)
	transcriber.Close()
	// Both tracks were created; closing must not panic or block.
}

// TestTranscriberFailThenEndTrackDoesNotPanic pins the daemon-crashing race
// found in the Gate A E2E: a provider failure (failTrack closes done) racing
// a track-end notification (endTrack also closed done) used to double-close
// the channel and panic the whole process.
func TestTranscriberFailThenEndTrackDoesNotPanic(t *testing.T) {
	provider := &failingSttProvider{}
	transcriber := NewTranscriber(provider, func(AttributedSttEvent) {})
	source := AudioSource{ParticipantID: "h", ParticipantName: "H", TrackName: "mic"}
	transcriber.AcceptAudio(source, opusFrame())
	time.Sleep(50 * time.Millisecond) // let the pump fail
	// TrackEnded for an already-failed track must be a silent no-op.
	transcriber.TrackEnded(source)
	transcriber.TrackEnded(source)
	transcriber.Close()
}

type failingSttProvider struct{}

func (*failingSttProvider) CreateSession(frame AudioFrame) (StreamingSttSession, error) {
	return nil, errTranscriberFailure
}

var errTranscriberFailure = &transcriberFailure{}

type transcriberFailure struct{}

func (*transcriberFailure) Error() string { return "provider failed" }
